File: src/commands/backup.rs
use serde::Serialize;
use serde_json::Value;

use crate::api::{ApiClient, ApiError};
use crate::output::{
    OutputFormat, extract_pagination, format_archivable_type, format_option, print_json,
    print_key_value, print_message, print_pagination, print_table,
};

#[derive(Debug, Serialize)]
struct ListBackupsQuery {
    #[serde(skip_serializing_if = "Option::is_none")]
    r#type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    site_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    environment_id: Option<String>,
    page: u32,
    per_page: u32,
}

#[derive(Debug, Serialize)]
struct CreateBackupRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    site_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    environment_id: Option<String>,
    scope: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
}

pub fn list(
    client: &ApiClient,
    site_id: Option<String>,
    environment_id: Option<String>,
    backup_type: Option<String>,
    page: u32,
    per_page: u32,
    format: OutputFormat,
) -> Result<(), ApiError> {
    let query = ListBackupsQuery {
        r#type: backup_type,
        site_id,
        environment_id,
        page,
        per_page,
    };
    let response: Value = client.get_with_query("/api/v1/vector/backups", &query)?;

    if format == OutputFormat::Json {
        print_json(&response);
        return Ok(());
    }

    let backups = response["data"]
        .as_array()
        .ok_or_else(|| ApiError::Other("Invalid response format".to_string()))?;

    if backups.is_empty() {
        print_message("No backups found.");
        return Ok(());
    }

    let rows: Vec<Vec<String>> = backups
        .iter()
        .map(|b| {
            vec![
                b["id"].as_str().unwrap_or("-").to_string(),
                b["archivable_type"]
                    .as_str()
                    .map(format_archivable_type)
                    .unwrap_or_else(|| "-".to_string()),
                b["type"].as_str().unwrap_or("-").to_string(),
                b["scope"].as_str().unwrap_or("-").to_string(),
                b["status"].as_str().unwrap_or("-").to_string(),
                format_option(&b["description"].as_str().map(String::from)),
                format_option(&b["created_at"].as_str().map(String::from)),
            ]
        })
        .collect();

    print_table(
        vec![
            "ID",
            "Model",
            "Type",
            "Scope",
            "Status",
            "Description",
            "Created",
        ],
        rows,
    );

    if let Some((current, last, total)) = extract_pagination(&response) {
        print_pagination(current, last, total);
    }

    Ok(())
}

pub fn show(client: &ApiClient, backup_id: &str, format: OutputFormat) -> Result<(), ApiError> {
    let response: Value = client.get(&format!("/api/v1/vector/backups/{}", backup_id))?;

    if format == OutputFormat::Json {
        print_json(&response);
        return Ok(());
    }

    let backup = &response["data"];

    print_key_value(vec![
        ("ID", backup["id"].as_str().unwrap_or("-").to_string()),
        (
            "Model",
            backup["archivable_type"]
                .as_str()
                .map(format_archivable_type)
                .unwrap_or_else(|| "-".to_string()),
        ),
        (
            "Model ID",
            backup["archivable_id"].as_str().unwrap_or("-").to_string(),
        ),
        ("Type", backup["type"].as_str().unwrap_or("-").to_string()),
        ("Scope", backup["scope"].as_str().unwrap_or("-").to_string()),
        (
            "Status",
            backup["status"].as_str().unwrap_or("-").to_string(),
        ),
        (
            "Description",
            format_option(&backup["description"].as_str().map(String::from)),
        ),
        (
            "File Snapshot ID",
            format_option(&backup["file_snapshot_id"].as_str().map(String::from)),
        ),
        (
            "Database Snapshot ID",
            format_option(&backup["database_snapshot_id"].as_str().map(String::from)),
        ),
        (
            "Started At",
            format_option(&backup["started_at"].as_str().map(String::from)),
        ),
        (
            "Completed At",
            format_option(&backup["completed_at"].as_str().map(String::from)),
        ),
        (
            "Created At",
            format_option(&backup["created_at"].as_str().map(String::from)),
        ),
        (
            "Updated At",
            format_option(&backup["updated_at"].as_str().map(String::from)),
        ),
    ]);

    Ok(())
}

pub fn create(
    client: &ApiClient,
    site_id: Option<String>,
    environment_id: Option<String>,
    scope: &str,
    description: Option<String>,
    format: OutputFormat,
) -> Result<(), ApiError> {
    if site_id.is_none() && environment_id.is_none() {
        return Err(ApiError::Other(
            "Either --site-id or --environment-id is required".to_string(),
        ));
    }

    let body = CreateBackupRequest {
        site_id,
        environment_id,
        scope: scope.to_string(),
        description,
    };

    let response: Value = client.post("/api/v1/vector/backups", &body)?;

    if format == OutputFormat::Json {
        print_json(&response);
        return Ok(());
    }

    let backup = &response["data"];
    print_message(&format!(
        "Backup created: {} ({})",
        backup["id"].as_str().unwrap_or("-"),
        backup["status"].as_str().unwrap_or("-")
    ));

    print_key_value(vec![
        ("ID", backup["id"].as_str().unwrap_or("-").to_string()),
        ("Type", backup["type"].as_str().unwrap_or("-").to_string()),
        (
            "Status",
            backup["status"].as_str().unwrap_or("-").to_string(),
        ),
        (
            "Description",
            format_option(&backup["description"].as_str().map(String::from)),
        ),
        (
            "Created At",
            format_option(&backup["created_at"].as_str().map(String::from)),
        ),
    ]);

    Ok(())
}

pub fn download_create(
    client: &ApiClient,
    backup_id: &str,
    format: OutputFormat,
) -> Result<(), ApiError> {
    let response: Value =
        client.post_empty(&format!("/api/v1/vector/backups/{}/downloads", backup_id))?;

    if format == OutputFormat::Json {
        print_json(&response);
        return Ok(());
    }

    let data = &response["data"];
    print_message(&format!(
        "Download requested: {} ({})",
        data["id"].as_str().unwrap_or("-"),
        data["status"].as_str().unwrap_or("-")
    ));
    print_message("\nCheck status with:");
    print_message(&format!(
        "  vector backup download status {} {}",
        backup_id,
        data["id"].as_str().unwrap_or("DOWNLOAD_ID")
    ));

    Ok(())
}

pub fn download_status(
    client: &ApiClient,
    backup_id: &str,
    download_id: &str,
    format: OutputFormat,
) -> Result<(), ApiError> {
    let response: Value = client.get(&format!(
        "/api/v1/vector/backups/{}/downloads/{}",
        backup_id, download_id
    ))?;

    if format == OutputFormat::Json {
        print_json(&response);
        return Ok(());
    }

    let data = &response["data"];
    print_key_value(vec![
        ("ID", data["id"].as_str().unwrap_or("-").to_string()),
        ("Status", data["status"].as_str().unwrap_or("-").to_string()),
        (
            "Size (bytes)",
            format_option(&data["size_bytes"].as_u64().map(|v| v.to_string())),
        ),
        (
            "Duration (ms)",
            format_option(&data["duration_ms"].as_u64().map(|v| v.to_string())),
        ),
        (
            "Error",
            format_option(&data["error_message"].as_str().map(String::from)),
        ),
        (
            "Download URL",
            format_option(&data["download_url"].as_str().map(String::from)),
        ),
        (
            "Download Expires",
            format_option(&data["download_expires_at"].as_str().map(String::from)),
        ),
        (
            "Started At",
            format_option(&data["started_at"].as_str().map(String::from)),
        ),
        (
            "Completed At",
            format_option(&data["completed_at"].as_str().map(String::from)),
        ),
        (
            "Created At",
            format_option(&data["created_at"].as_str().map(String::from)),
        ),
    ]);

    Ok(())
}

File: internal/commands/backup.go
package commands

import (
	"encoding/json"
	"fmt"
	"io"
	"strings"

	"github.com/spf13/cobra"

	"github.com/built-fast/vector-cli/internal/output"
)

const backupsBasePath = "/api/v1/vector/backups"

// NewBackupCmd creates the backup command group.
func NewBackupCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "backup",
		Short: "Manage backups",
		Long:  "Manage backups to protect your site data.",
	}

	cmd.AddCommand(newBackupListCmd())
	cmd.AddCommand(newBackupShowCmd())
	cmd.AddCommand(newBackupCreateCmd())
	cmd.AddCommand(NewBackupDownloadCmd())

	return cmd
}

// formatArchivableType formats the archivable_type for display.
// e.g., "vector_site" becomes "Site", "vector_environment" becomes "Environment".
func formatArchivableType(raw string) string {
	raw = strings.TrimPrefix(raw, "vector_")
	if raw == "" {
		return "-"
	}
	// Capitalize first letter
	return strings.ToUpper(raw[:1]) + raw[1:]
}

func newBackupListCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List backups",
		Long:  "Retrieve a paginated list of backups, optionally filtered by type, site, or environment.",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			app, err := requireApp(cmd)
			if err != nil {
				return err
			}

			page, perPage := getPagination(cmd)
			query := buildPaginationQuery(page, perPage)

			if cmd.Flags().Changed("site-id") {
				v, _ := cmd.Flags().GetString("site-id")
				if v != "" {
					query.Set("site_id", v)
				}
			}
			if cmd.Flags().Changed("environment-id") {
				v, _ := cmd.Flags().GetString("environment-id")
				if v != "" {
					query.Set("environment_id", v)
				}
			}
			if cmd.Flags().Changed("type") {
				v, _ := cmd.Flags().GetString("type")
				if v != "" {
					query.Set("type", v)
				}
			}

			resp, err := app.Client.Get(cmd.Context(), backupsBasePath, query)
			if err != nil {
				return fmt.Errorf("failed to list backups: %w", err)
			}
			defer func() { _ = resp.Body.Close() }()

			body, err := io.ReadAll(resp.Body)
			if err != nil {
				return fmt.Errorf("failed to list backups: %w", err)
			}

			if app.Format == output.JSON {
				data, err := parseResponseData(body)
				if err != nil {
					return fmt.Errorf("failed to list backups: %w", err)
				}
				return output.PrintJSON(cmd.OutOrStdout(), json.RawMessage(data))
			}

			data, meta, err := parseResponseWithMeta(body)
			if err != nil {
				return fmt.Errorf("failed to list backups: %w", err)
			}

			var items []map[string]any
			if err := json.Unmarshal(data, &items); err != nil {
				return fmt.Errorf("failed to list backups: %w", err)
			}

			headers := []string{"ID", "MODEL", "TYPE", "SCOPE", "STATUS", "DESCRIPTION", "CREATED"}
			var rows [][]string
			for _, item := range items {
				rows = append(rows, []string{
					getString(item, "id"),
					formatArchivableType(getString(item, "archivable_type")),
					getString(item, "type"),
					getString(item, "scope"),
					getString(item, "status"),
					formatString(getString(item, "description")),
					getString(item, "created_at"),
				})
			}

			output.PrintTable(cmd.OutOrStdout(), headers, rows)
			printPaginationIfNeeded(cmd.OutOrStdout(), meta)
			return nil
		},
	}
	addPaginationFlags(cmd)
	cmd.Flags().String("site-id", "", "Filter by site ID")
	cmd.Flags().String("environment-id", "", "Filter by environment ID")
	cmd.Flags().String("type", "", "Filter by backup type (site/environment)")
	return cmd
}

func newBackupShowCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "show <id>",
		Short: "Show a backup",
		Long:  "Display details of a specific backup.",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			app, err := requireApp(cmd)
			if err != nil {
				return err
			}

			resp, err := app.Client.Get(cmd.Context(), backupsBasePath+"/"+args[0], nil)
			if err != nil {
				return fmt.Errorf("failed to get backup: %w", err)
			}
			defer func() { _ = resp.Body.Close() }()

			body, err := io.ReadAll(resp.Body)
			if err != nil {
				return fmt.Errorf("failed to get backup: %w", err)
			}

			data, err := parseResponseData(body)
			if err != nil {
				return fmt.Errorf("failed to get backup: %w", err)
			}

			if app.Format == output.JSON {
				return output.PrintJSON(cmd.OutOrStdout(), json.RawMessage(data))
			}

			var item map[string]any
			if err := json.Unmarshal(data, &item); err != nil {
				return fmt.Errorf("failed to get backup: %w", err)
			}

			output.PrintKeyValue(cmd.OutOrStdout(), []output.KeyValue{
				{Key: "ID", Value: getString(item, "id")},
				{Key: "Model", Value: formatArchivableType(getString(item, "archivable_type"))},
				{Key: "Model ID", Value: getString(item, "archivable_id")},
				{Key: "Type", Value: getString(item, "type")},
				{Key: "Scope", Value: getString(item, "scope")},
				{Key: "Status", Value: getString(item, "status")},
				{Key: "Description", Value: formatString(getString(item, "description"))},
				{Key: "File Snapshot ID", Value: formatString(getString(item, "file_snapshot_id"))},
				{Key: "Database Snapshot ID", Value: formatString(getString(item, "database_snapshot_id"))},
				{Key: "Started At", Value: formatString(getString(item, "started_at"))},
				{Key: "Completed At", Value: formatString(getString(item, "completed_at"))},
				{Key: "Created At", Value: getString(item, "created_at")},
				{Key: "Updated At", Value: getString(item, "updated_at")},
			})
			return nil
		},
	}
}

func newBackupCreateCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "create",
		Short: "Create a backup",
		Long:  "Create a new backup for a site or environment.",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			app, err := requireApp(cmd)
			if err != nil {
				return err
			}

			siteID, _ := cmd.Flags().GetString("site-id")
			envID, _ := cmd.Flags().GetString("environment-id")

			if siteID == "" && envID == "" {
				return fmt.Errorf("either --site-id or --environment-id is required")
			}

			scope, _ := cmd.Flags().GetString("scope")
			description, _ := cmd.Flags().GetString("description")

			reqBody := map[string]any{
				"type":  "manual",
				"scope": scope,
			}

			if siteID != "" {
				reqBody["site_id"] = siteID
			}
			if envID != "" {
				reqBody["environment_id"] = envID
			}
			if description != "" {
				reqBody["description"] = description
			}

			resp, err := app.Client.Post(cmd.Context(), backupsBasePath, reqBody)
			if err != nil {
				return fmt.Errorf("failed to create backup: %w", err)
			}
			defer func() { _ = resp.Body.Close() }()

			body, err := io.ReadAll(resp.Body)
			if err != nil {
				return fmt.Errorf("failed to create backup: %w", err)
			}

			data, err := parseResponseData(body)
			if err != nil {
				return fmt.Errorf("failed to create backup: %w", err)
			}

			if app.Format == output.JSON {
				return output.PrintJSON(cmd.OutOrStdout(), json.RawMessage(data))
			}

			var item map[string]any
			if err := json.Unmarshal(data, &item); err != nil {
				return fmt.Errorf("failed to create backup: %w", err)
			}

			w := cmd.OutOrStdout()
			output.PrintMessage(w, fmt.Sprintf("Backup created: %s (%s)", getString(item, "id"), getString(item, "status")))
			output.PrintMessage(w, "")

			output.PrintKeyValue(w, []output.KeyValue{
				{Key: "ID", Value: getString(item, "id")},
				{Key: "Model", Value: formatArchivableType(getString(item, "archivable_type"))},
				{Key: "Model ID", Value: getString(item, "archivable_id")},
				{Key: "Type", Value: getString(item, "type")},
				{Key: "Scope", Value: getString(item, "scope")},
				{Key: "Status", Value: getString(item, "status")},
				{Key: "Description", Value: formatString(getString(item, "description"))},
				{Key: "File Snapshot ID", Value: formatString(getString(item, "file_snapshot_id"))},
				{Key: "Database Snapshot ID", Value: formatString(getString(item, "database_snapshot_id"))},
				{Key: "Started At", Value: formatString(getString(item, "started_at"))},
				{Key: "Completed At", Value: formatString(getString(item, "completed_at"))},
				{Key: "Created At", Value: getString(item, "created_at")},
				{Key: "Updated At", Value: getString(item, "updated_at")},
			})
			return nil
		},
	}

	cmd.Flags().String("site-id", "", "Site ID to back up")
	cmd.Flags().String("environment-id", "", "Environment ID to back up")
	cmd.Flags().String("scope", "full", "Backup scope (full/database/files)")
	cmd.Flags().String("description", "", "Description for the backup")

	return cmd
}
